### ISSUE TITLE
Rule "no-invalid-this" produces false positives.

### ISSUE THREAD

**Issue by PEConn** [eslint/eslint#13894]
Node version: v12.14.1
npm version: v6.14.8
Local ESLint version: v6.8.0 (Currently used)
Global ESLint version: Not found

I'm using `@typescript-eslint/parser`.

My config:

<details>
<summary>Configuration</summary>

```js
{
  "extends": "google",
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module"
  },
  "rules": {
    "max-len": [2, 100, {
      "ignoreComments": true,
      "ignoreUrls": true,
      "tabWidth": 2
    }],
    "no-implicit-coercion": [2, {
      "boolean": false,
      "number": true,
      "string": true
    }],
    "no-unused-expressions": [2, {
      "allowShortCircuit": true,
      "allowTernary": false
    }],
    "no-unused-vars": [2, {
      "vars": "all",
      "args": "after-used",
      "argsIgnorePattern": "(^reject$|^_$)",
      "varsIgnorePattern": "(^_$)"
    }],
    "quotes": [2, "single"],
    "require-jsdoc": 0,
    "valid-jsdoc": 0,
    "prefer-arrow-callback": 1,
    "no-var": 1
  },
  "env": {
    "node": true,
    "jasmine": true
  },
  "overrides": [
    {
      "files": ["**/*.ts"],
      "extends": [
        "plugin:@typescript-eslint/eslint-recommended",
        "plugin:@typescript-eslint/recommended"
      ],
      "rules": {
        "@typescript-eslint/no-non-null-assertion": 0
      }
    }
  ]
}

```

</details>

What I did:

```js
enum LogLevel { Debug, Info, Warn, Error }

/** Data class to store level and message of pending */
class PendingLog {
  constructor(public level: LogLevel, public message: string) {}
}

/**
 * A Log that wraps another Log, saving up all the log calls to be applied when flush is called.
 *
 * It doesn't currently support arguments to the log messages.
 */
export class BufferedLog implements Log {
  private pendingLogs: Array<PendingLog> = [];

  constructor(private innerLog: Log) {}

  /* eslint-disable no-invalid-this */
  debug = this.addLogFunction(LogLevel.Debug);
  info = this.addLogFunction(LogLevel.Info);
  warn = this.addLogFunction(LogLevel.Warn);
  error = this.addLogFunction(LogLevel.Error);
  /* eslint-enable no-invalid-this */

  /** Creates a function that adds a log at the given level. */
  private addLogFunction(level: LogLevel): (message: string) => void {
    return (message: string): void => {
      this.pendingLogs.push(new PendingLog(level, message));
    };
  }

  // ...
}

```

```bash
eslint "src/**/*.{js,ts}"
```

**What did you expect to happen?**

I believe my use of `this` between the `eslint-disable` comments is valid. (I could be wrong.)

**What actually happened? Please include the actual, raw output from ESLint.**

Without `/* eslint-disable no-invalid-this */` I get `error  Unexpected 'this'  no-invalid-this` 4 times for those 4 lines.

**Are you willing to submit a pull request to fix this bug?**

Maybe - it very much depends how difficult it is to fix.

You could get a bit more context from [this PR][1] where I encountered the issue.

[1]: https://github.com/GoogleChromeLabs/bubblewrap/pull/417/commits/5b8e06ef0b4f8d08792b90f676431826dc79a5b3

**Comment by mdjermanovic** [comment]
Hi @PEConn, thanks for the issue!

> I believe my use of this between the eslint-disable comments is valid. (I could be wrong.)

It looks valid to me, but the core `no-invalid-this` rule doesn't support class fields yet, since it's a [stage 3](https://github.com/tc39/proposal-class-fields) feature.

Can you try with [@typescript-eslint/no-invalid-this](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-invalid-this.md) instead of the core rule:

```json
{ 
  "no-invalid-this": "off",
  "@typescript-eslint/no-invalid-this": ["error"]
}
```

Since you're using ESLint with typescript, the following link will be probably useful:

https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code

**Comment by PEConn** [comment]
Thanks for the advice - I changed by rules as you suggested, but `"@typescript-eslint/no-invalid-this"` also triggers. I raised an [issue with typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/issues/2840).
